### PR TITLE
[New Rule] Adds support for enforcing JSDoc

### PR DESCRIPTION
**Pull request by orta**
#### PR checklist

- [x] Addresses an existing issue: #623
- [x] New feature
  - [x] Includes tests
- [ ] Documentation update

#### What changes did you make?

Hey there, I had been wanting to add a [feature](https://github.com/eslint/eslint/pull/7867) to ESLint for a while, but we moved to TypeScript, and so I actually needed it in TSLint also. 

This is close to an exact port of the "require-jsdoc" port from ESLint, but with less rigid defaults. It will default to only raising errors to `export`ed classes/functions. You can make it apply to all functions though, too.

#### Is there anything you'd like reviewers to focus on?

- I've only done JavaScript for ~6 months TS for 1 month, there are things I've probably not taken into account.
- Should this take into account TS' private / public functions?
- Not sure what I have to do re:docs

--

Sidenotes:

* I updated the settings for VS Code - the JSON shape has changed, and it would show the JS files instead of just TS files in search
* I added a quick feature to the ruleTestRunner so you could run a subset of tests with ` node ./build/test/ruleTestRunner.js "test/rules/require-jsdoc/*/tslint.json"` - 30 seconds  was too long.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @orta! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/rules/requireJsdocRule.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    private getRequireJSDocOptions() {
+        const DEFAULT_OPTIONS = {
+            classDeclaration: true,
+            exportedOnly: true,
+            functionDeclaration: true,
+            methodDefinition: true,
+        };
+
+        const ruleOptions = this.getOptions();
+        return Object.assign(DEFAULT_OPTIONS, ruleOptions && ruleOptions[0]);
```
Comment (ajafff) [inline src/rules/requireJsdocRule.ts]: you could do this once in the constructor and assign the result to a property

File: src/rules/requireJsdocRule.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                if (settings.exportedOnly && !this.isExportedNode(node)) {
+                    return;
+                }
+                this.addFailureAtNode(node.name, Rule.FAILURE_STRING);
+            }
+        }
+    }
+
+    private isExportedNode(node: ts.Node): boolean {
+        if (Lint.hasModifier(node.modifiers, ts.SyntaxKind.ExportKeyword)) { return true; }
```
Comment (ajafff) [inline src/rules/requireJsdocRule.ts]: that doesn't seem to be the correct way to check for exports.
consider the following code:
```ts
export function exportedFunction() { // <-- JSDoc required
    function nestedFunction() {} // <-- JSDoc required
}
```
maybe a better approach would be
```ts
// if you use a discriminated union as parameter this switch will be exhaustive
switch(node.kind) {
    case ts.SyntaxKind.FunctionDeclaration:
    case ts.SyntaxKind.ClassDeclaration:
        return Lint.hasModifier(node.modifiers, ts.SyntaxKind.ExportKeyword);
    case ts.SyntaxKind.MethodDeclaration:
        return Lint.hasModifier(node.parent!.modifiers, ts.SyntaxKind.ExportKeyword);
    case ts.SyntaxKind.FunctionExpression:
        return node.parent!.kind === ts.SyntaxKind.VariableDeclaration && Lint.hasModifier(node.parent!.parent!.parent!.modifiers, ts.SyntaxKind.ExportKeyword);
        // .parent!.parent! is the VariableDeclarationList, .parent!.parent!.parent! is the VariableStatement, which can be exported
}
```

File: src/rules/requireJsdocRule.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    private checkJSDoc(node: ts.FunctionDeclaration | ts.FunctionExpression | ts.ClassDeclaration | ts.MethodDeclaration) {
+        if (node.name) {
+            const checker = this.getTypeChecker();
+            const symbol = checker.getSymbolAtLocation(node.name);
+            const hasDocs = symbol.getDocumentationComment().length > 0;
+
+            if (!hasDocs ) {
+                const settings = this.getRequireJSDocOptions();
+                // console.log(settings);
+                if (settings.exportedOnly && !this.isExportedNode(node)) {
```
Comment (ajafff) [inline src/rules/requireJsdocRule.ts]: you should check if you want to enforce JSDoc on this node *before* you use the typechecker

File: test/ruleTestRunner.ts [hunk 4]
```diff
@@ -25,7 +25,12 @@ console.log();
 console.log(colors.underline("Testing Lint Rules:"));
 /* tslint:enable:no-console */
 
-const testDirectories = glob.sync("test/rules/**/tslint.json").map(path.dirname);
+// Support running a subset of lint rules:
```
Comment (ajafff) [inline test/ruleTestRunner.ts]: is this change necessary?

you can run rule tests by simply using `./bin/tslint --test some/glob*pattern`

File: test/rules/require-jsdoc/exports-only-classes-only/test.ts.lint [hunk 5]
```diff
@@ -0,0 +1,17 @@
+/** I am documented */
+export class Documented {
+      
+}
+
+export class Undocumented {
+             ~~~~~~~~~~~~   [Missing JSDoc comment.]
```
Comment (ajafff) [inline test/rules/require-jsdoc/exports-only-classes-only/test.ts.lint]: please add some tests for nested functions

File: src/rules/requireJsdocRule.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "require-jsdoc",
+        description: "Require JSDoc comments.",
+        hasFix: false,
+        optionsDescription: Lint.Utils.dedent`
+            Ensure your project is documented. Defaults to only show errors 
+            on undocumented classes or functions which are exported.
+
+            You can opt out of checking for a specific set of declarations in
+            the options.
```
Comment (ajafff) [inline src/rules/requireJsdocRule.ts]: this should describe how to use the `options` object

File: src/rules/requireJsdocRule.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                methodDefinition: {
+                    type: "boolean",
+                },
+            },
+            additionalProperties: false,
+        },
+        optionExamples: ['[true, {"exportedOnly": false, "classDeclaration": false}]',
+                         '[true, {"exportedOnly": false, "classDeclaration": false, functionDeclaration: false, methodDefinition: false}]'],
+
+        type: "style",
```
Comment (ajafff) [inline src/rules/requireJsdocRule.ts]: requiresTypeInfo: true

File: test/ruleTestRunner.ts (hunk 4 above)
Comment (orta) [inline test/ruleTestRunner.ts]: Ah yeah, I did not know about this feature! useful 💯

**Comment by nchen63** [comment]
I think this duplicates behavior from `completed-docs` https://github.com/palantir/tslint/pull/2095

**Comment by orta** [comment]
Oh yeah -  @nchen63 it does, and that's much more comprehensive 👍 

@ajafff - thanks for the review ( it was useful, and I got pretty far with your changes, but not much point continuing ) 

I'm going to close this. :1st_place_medal:
